### PR TITLE
Progress Bar Rewrite

### LINKED ISSUE #2992: spinners aren't contained in progres bar on bottom of the screen
in places where spinners should be seen on progres bar there's a blank space
good visible in start of this beatmap: https://osu.ppy.sh/beatmapsets/323961#osu/719902
logs:
[database.log](https://github.com/ppy/osu/files/2181325/database.log)
[network.log](https://github.com/ppy/osu/files/2181326/network.log)
[performance.log](https://github.com/ppy/osu/files/2181327/performance.log)
[performance-gamethread.audio.log](https://github.com/ppy/osu/files/2181328/performance-gamethread.audio.log)
[performance-gamethread.draw.log](https://github.com/ppy/osu/files/2181329/performance-gamethread.draw.log)
[performance-gamethread.input.log](https://github.com/ppy/osu/files/2181330/performance-gamethread.input.log)
[performance-gamethread.update.log](https://github.com/ppy/osu/files/2181331/performance-gamethread.update.log)
[runtime.log](https://github.com/ppy/osu/files/2181332/runtime.log)

### PR DESCRIPTION
**Pull request by ticotaco72**
Rewrite of progress bar graph. It isn't perfect (i can't find better solution), but it's still better than original. 

Some screenshots:

- (partially) Closes #2992 

---

Makes progress bar more accurate to difficulty.
haha

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 1]
```diff
@@ -23,6 +24,65 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+                new Aim(),
+                new Speed()
+            };
+
+            double sectionLength = section_length * timeRate;
+
+            // The first object doesn't generate a strain, so we begin with an incremented section end
+            double currentSectionEnd = 2 * sectionLength;
+
+            foreach ( OsuDifficultyHitObject h in difficultyBeatmap )
```
Comment (AlFasGD) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Whitespace style doesn't comply with project code style, remove spaces between parentheses, prefer usage of `var`

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 2]
```diff
@@ -23,6 +24,65 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+            {
+                while (h.BaseObject.StartTime > currentSectionEnd)
+                {
+                    foreach (Skill s in skills)
```
Comment (AlFasGD) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Prefer usage of `var`

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 3]
```diff
@@ -23,6 +24,65 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+                    {
+                        s.SaveCurrentPeak();
+                        s.StartNewSectionFrom(currentSectionEnd);
+                    }
+
+                    currentSectionEnd += sectionLength;
+                }
+
+                foreach (Skill s in skills)
```
Comment (AlFasGD) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Prefer usage of `var`

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 4]
```diff
@@ -23,6 +24,65 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+                    s.Process(h);
+            }
+
+            var aimRating = new List<double>();
+            var speedRating = new List<double>();
+            var osuDifficultySectionRating = new List<double>();
+
+            foreach(double x in skills[0].StrainPeaks)
```
Comment (AlFasGD) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Add space between `foreach` and parentheses

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 5]
```diff
@@ -23,6 +24,65 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+            {
```
Comment (AlFasGD) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Brackets can be safely removed

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 6]
```diff
@@ -87,4 +147,4 @@ protected override DifficultyAttributes Calculate(IBeatmap beatmap, Mod[] mods,
             new OsuModHardRock(),
         };
     }
-}
+}
```
Comment (AlFasGD) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Don't remove extra ending line as created by IDEs by default

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs [hunk 7]
```diff
@@ -33,6 +33,37 @@ public TaikoDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+
+            if (!calculateStrainValues(difficultyHitObjects, timeRate))
+                return new List<double>();
+
+            var highestStrains = calculateDifficulty(difficultyHitObjects, timeRate);
+            var taikoDifficultySectionRating = new List<double>();
+            var startAt = difficultyHitObjects[0].BaseHitObject.StartTime / (strain_step * timeRate);
+
+            for (int x = 0; x < highestStrains.Count; x++)
+            {
```
Comment (AlFasGD) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs]: Remove brackets

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 8]
```diff
@@ -78,9 +78,14 @@ public HUDOverlay(ScoreProcessor scoreProcessor, RulesetContainer rulesetContain
 
             BindProcessor(scoreProcessor);
             BindRulesetContainer(rulesetContainer);
-
-            Progress.Objects = rulesetContainer.Objects;
             Progress.AudioClock = offsetClock;
+            var difficultyCalculator = rulesetContainer.Ruleset.CreateDifficultyCalculator(working);
+            Progress.StrainStep = difficultyCalculator.StrainStep();
+            if (rulesetContainer.Ruleset.LegacyID==0)
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/HUDOverlay.cs]: Use spaces between operator and operands

File: osu.Game/Screens/Play/SongProgress.cs [hunk 9]
```diff
@@ -57,6 +60,24 @@ public IEnumerable<HitObject> Objects
             }
         }
 
+        private double strainStep;
+
+        public double StrainStep
+        {
+            get
+            {return strainStep;}
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgress.cs]: Either use lambdas, or put brackets in the same lines with the `get`/`set` but with spaces between brackets

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 10]
```diff
@@ -6,43 +6,207 @@
...
     public class SongProgressGraph : SquareGraph
     {
+        private List<double> strains = new List<double>();
+
+        public List<double> Strains
+        {
+            get { return strains; }
+            set
+            {
+                for(int x = 0; x < value.Count; x++)
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: Use space between `for` and parenthesis

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 11]
```diff
@@ -6,43 +6,207 @@
...
+                        strains.Add(value[x]);
+                        strains.Add(value[x]);
+                        Values.Add(value[x]);
+                        Values.Add(value[x]);
+                    }
+                    else
+                    {
+                        strains.Add(value[x]);
+                        Values.Add(value[x]);
+                    }
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: The `else` code block can be moved outside and only add the values again if `x == 0`.

File: osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs [hunk 12]
```diff
@@ -103,5 +103,25 @@ IEnumerable<Mod> createDifficultyAdjustmentModCombinations(IEnumerable<Mod> curr
         /// <param name="timeRate">The rate of time in <paramref name="beatmap"/>.</param>
         /// <returns>A structure containing the difficulty attributes.</returns>
         protected abstract DifficultyAttributes Calculate(IBeatmap beatmap, Mod[] mods, double timeRate);
+
+        public List<double> DifficultySectionRating ()
```
Comment (AlFasGD) [inline osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs]: Remove space between function name and parentheses

File: osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs [hunk 13]
```diff
@@ -103,5 +103,25 @@ IEnumerable<Mod> createDifficultyAdjustmentModCombinations(IEnumerable<Mod> curr
...
+        {
+            IBeatmap playableBeatmap = beatmap.GetPlayableBeatmap(ruleset.RulesetInfo);
+            var clock = new StopwatchClock();
+
+            return DifficultySectionRating (playableBeatmap, clock.Rate);
```
Comment (AlFasGD) [inline osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs]: Same here

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 14]
```diff
@@ -6,43 +6,207 @@
...
                 objects = value;
 
-                const int granularity = 200;
-                Values = new int[granularity];
-
                 if (!objects.Any())
-                    return;
+                {
+                    for (int x = 0; x < Strains.Count; x++)
+                    {
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: Remove brackets, can also use a `foreach` loop

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 15]
```diff
@@ -6,43 +6,207 @@
...
+                        Strains[x] = 0;
+                    }
+                }
+                else
+                {
+                    if (Strains.Count==0)
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: Prefer usage of `Any()` here too

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 16]
```diff
@@ -6,43 +6,207 @@
...
+                }
 
-                var firstHit = objects.First().StartTime;
-                var lastHit = objects.Max(o => (o as IHasEndTime)?.EndTime ?? o.StartTime);
+                var values = new List<int>();
 
-                if (lastHit == 0)
-                    lastHit = objects.Last().StartTime;
+                for (int x = 0; x < strains.Count; x++)
+                {
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: Remove brackets, apply for following occurrences as well

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 17]
```diff
@@ -6,43 +6,207 @@
...
+                    values.Add(0);
+                }
 
-                var interval = (lastHit - firstHit + 1) / granularity;
+                var startOfLists = (objects.First().StartTime - objects.First().StartTime % (strainStep * audioClock.Rate))/(strainStep * audioClock.Rate);
+                if (strainStep==1)
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: Add spaces between operator and operands, apply for the following code, too many occurrences

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 18]
```diff
@@ -6,43 +6,207 @@
...
+                                    }
+                                    else
+                                        continue;
+                                }
+                                else
+                                    continue;
+                            }
+
+                            double endOfTouch = -1;
+                            if ( (x+1) * interval + startOfLists >= breaks[i].StartTime && (x+1) * interval + startOfLists < breaks[i].EndTime)
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: Simplify expression by storing `(x + 1) * interval + startOfLists` in a local constant

File: osu.Game/Screens/Play/SongProgressGraph.cs [hunk 19]
```diff
@@ -6,43 +6,207 @@
...
+                if (interval==1)
+                {
+                    strainStep = 400;
+                }
+
+                for (int x = 0; x < strains.Count; x++)
+                {
+                    for (int i = 0; i < breaks.Count; i++)
+                    {
+                        if ( breaks[i].Duration >= interval / 2)
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SongProgressGraph.cs]: Remove spaces between parentheses, apply for following occurrences as well

File: osu.Game/Screens/Play/SquareGraph.cs [hunk 20]
```diff
@@ -112,25 +112,32 @@ private void redrawFilled()
         /// </summary>
         private void recalculateValues()
         {
-            var newValues = new List<float>();
+            var newValues = new List<double>();
 
-            if (values == null)
+            if (values.Count == 0)
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SquareGraph.cs]: Prefer usage of `Any()`

File: osu.Game/Screens/Play/SquareGraph.cs [hunk 21]
```diff
@@ -112,25 +112,32 @@ private void redrawFilled()
...
-            for (float i = 0; i < values.Length; i += step)
+            float step = values.Count / (float)ColumnCount;
+            for (float i = 0; i < values.Count; i += step)
             {
-                newValues.Add((float)values[(int)i] / max);
+                double sum = 0;
+                float iteration = 0;
+                for (float x = i; x < i+step && x < values.Count; x++)
+                {
+                    sum += values[(int) x];
```
Comment (AlFasGD) [inline osu.Game/Screens/Play/SquareGraph.cs]: Remove space between cast and expression

File: osu.Game/Screens/Ranking/ResultsPageScore.cs [hunk 22]
```diff
@@ -31,7 +32,11 @@ public class ResultsPageScore : ResultsPage
     {
         private ScoreCounter scoreCounter;
 
-        public ResultsPageScore(Score score, WorkingBeatmap beatmap) : base(score, beatmap) { }
+        private readonly WorkingBeatmap working;
+
+        private readonly Score score;
+
+        public ResultsPageScore(Score score, WorkingBeatmap beatmap) : base(score, beatmap) { this.score=score; working = beatmap; }
```
Comment (AlFasGD) [inline osu.Game/Screens/Ranking/ResultsPageScore.cs]: Don't compact the constructor into one line

**Review by Rekkonnect — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs (hunk 2 above)
Comment (ticotaco72) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: you mean `foreach (var s in skills)` or making it for loop and then var?

Comment (AlFasGD) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: The first one

Comment (ticotaco72) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: ok thanks

File: osu.Game/Screens/Play/SongProgressGraph.cs (hunk 14 above)
Comment (ticotaco72) [reply osu.Game/Screens/Play/SongProgressGraph.cs]: can't be `foreach` because i can't assign value to foreach variable

Comment (AlFasGD) [reply osu.Game/Screens/Play/SongProgressGraph.cs]: Doesn't the following work?

```
foreach (var s in Strains)
    s = 0;
```

Comment (ticotaco72) [reply osu.Game/Screens/Play/SongProgressGraph.cs]: yeah, compiler error goes if like there

**Comment by ticotaco72** [comment]
i can't repair this two style problems, because then i got compiler error

**Comment by ticotaco72** [comment]
Can't determine what's wrong

**Comment by ticotaco72** [comment]
Closing as this needs some more work

**Pull request closed without merging**